src/drivers/fluid_sdl2.c

```diff
@@ -0,0 +1,250 @@
+
+        sample_size = sizeof(short);
+        write_ptr   = fluid_synth_write_s16;
+
+        aspec.format = AUDIO_S16SYS;
+    }
+    else
+    {
+        FLUID_LOG(FLUID_ERR, "Unhandled sample format");
+        return NULL;
+    }
+
+    /* Compile the format buffer */
+    aspec.channels   = 2;
+    aspec.samples    = aspec.channels * ((period_size + 7) & ~7);
+    aspec.callback   = (SDL_AudioCallback)SDLAudioCallback;
+
+    /* Start SDL library */
+    if (SDL_Init(SDL_INIT_AUDIO))
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to initialize SDL2: %s", SDL_GetError());
+        return NULL;
+    }
+
+    /* Set default device to use */
+    device   = NULL;
+    dev_name = NULL;
+
+    /* get the selected device name. if none is specified, use default device. */
+    if(fluid_settings_dupstr(settings, "audio.sdl2.device", &device) == FLUID_OK
+            && device != NULL && device[0] != '\0')
+    {
+        int n, nDevs = SDL_GetNumAudioDevices(0);
+
+        for (n = 0; n < nDevs; n++)
+        {
+            dev_name = SDL_GetAudioDeviceName(n, 0);
+
+            if(FLUID_STRCASECMP(dev_name, device) == 0)
+            {
+                FLUID_LOG(FLUID_DBG, "Selected audio device GUID: %s", dev_name);
+                break;
+            }
+        }
+
+        if (n >= nDevs)
+        {
+            FLUID_LOG(FLUID_DBG, "Audio device %s, using \"default\"", device);
+            dev_name = NULL;
+        }
+    }
+
+    if (device != NULL)
+    {
+        FLUID_FREE(device);
+    }
+
+    do {
+        /* create and clear the driver data */
+        dev = FLUID_NEW(fluid_sdl2_audio_driver_t);
+
+        if(dev == NULL)
+        {
+            FLUID_LOG(FLUID_ERR, "Out of memory");
+            break;
+        }
+
+        FLUID_MEMSET(dev, 0, sizeof(fluid_sdl2_audio_driver_t));
+
+        /* set device pointer to userdata */
+        aspec.userdata = dev;
+
+        /* Save copy of synth */
+        dev->synth = synth;
+
+        /* Save copy of other variables */
+        dev->write_ptr = write_ptr;
+        dev->frame_size = sample_size * aspec.channels;
+
+        /* Open audio device */
+        dev->devid = SDL_OpenAudioDevice(dev_name, 0, &aspec, &rspec, 0);
+
+        if (!dev->devid)
+        {
+            FLUID_LOG(FLUID_ERR, "Failed to open audio device");
+            break;
+        }
+
+        /* Start to play */
+        SDL_PauseAudioDevice(dev->devid, 0);
+
+        return (fluid_audio_driver_t *) dev;
+    } while (0);
+
+    delete_fluid_sdl2_audio_driver(&dev->driver);
+    return NULL;
+}
+
+
+void delete_fluid_sdl2_audio_driver(fluid_audio_driver_t *d)
+{
+    fluid_sdl2_audio_driver_t *dev = (fluid_sdl2_audio_driver_t *) d;
+
+    if (dev != NULL)
+    {
+        if (dev->devid)
+        {
+            /* Stop audio and close */
+            SDL_PauseAudioDevice(dev->devid, 1);
+            SDL_CloseAudioDevice(dev->devid);
+        }
+
+        FLUID_FREE(dev);
+    }
+
+    SDL_Quit();
```

Comment:
`fluid_get_audio_driver_name()` is redundant. The user knows which audio driver he gets, because he specified it with the setting `audio.driver`. Same is true for midi drivers.

Your proposal doesn't fix the problem that we've initialized sdl in `fluid_sdl2_audio_driver_settings()` and need to clean it up. No matter whether the user uses the sdl driver or not.

I propose to go even one step further and let the user call `SDL_Init()` and `SDL_Quit()` in his app. If sdl wasn't inited when `fluid_sdl2_audio_driver_settings()` is called, the sdl driver cannot (or should not) be used (i.e. undefined behaviour). Could you test how `SDL_GetNumAudioDevices(0)` behaves when not calling `SDL_Init()`?

Reply:
> fluid_get_audio_driver_name() is redundant. The user knows which audio driver he gets, because he specified it with the setting audio.driver. Same is true for midi drivers.

Ouch, you are right, I was thinking that `audio.driver` could be also set to `default`, but after looking the code, in reality it cannot be.

Actually, if we would like to move the calls to `SDL_Init()` and `SDL_Quit()` outside the library, we can use `SDL_WasInit()`, perhaps like this, in the place where `SDL_Init()` is called:

```
if (!SDL_WasInit(SDL_INIT_AUDIO))
{
        FLUID_LOG(FLUID_ERR, "SDL2 not initialized");
        return NULL;
}
```

and just remove the calls to `SDL_Quit()` where it exists.

BTW, I also opened a thread here:
https://discourse.libsdl.org/t/plugins-development-how-to-call-sdl-quit/25396
Perhaps some of SDL2 developers will have some advice to give us.

Reply:
`SDL_WasInit()` looks good! Let's follow this strategy and only register all the sdl fluid_settings if sdl was inited. And leave responibility for `SDL_Init()` and `_Quit()` to the application.

Reply:
After reading the source code of SDL2, I think that I got an idea.
Unlike SDL1, the calls to SDL2 are fully refcountered.
So:
1) calling `SDL_Init()` multiple times is not a problem, the number of calls is tracked.
2) instead of calling `SDL_Quit()`, we can call `SDL_QuitSubSystem()` to decrement the refcount increased by the previous `SDL_Init()`.
3) before quitting, it is responsibility of the application to call the final `SDL_Quit()` only one time.

In that way, the application is at least freed by the need to call `SDL_Init()`.